crates/ddl/src/core/compile/doc/mod.rs

```diff
@@ -0,0 +1,209 @@
+use codespan::FileId;
+use codespan_reporting::diagnostic::Diagnostic;
+use std::borrow::Cow;
+use std::io;
+use std::io::prelude::*;
+
+use crate::core;
+
+pub fn compile_module(
+    writer: &mut impl Write,
+    module: &core::Module,
+    report: &mut dyn FnMut(Diagnostic),
+) -> io::Result<()> {
+    let context = ModuleContext {
+        _file_id: module.file_id,
+    };
+
+    const MINIRESET: &str = include_str!("./minireset.min.css");
+    const STYLE: &str = include_str!("./style.css");
+
+    let pkg_name = env!("CARGO_PKG_NAME");
+    let pkg_version = env!("CARGO_PKG_VERSION");
+
+    write!(
+        writer,
+        "\
+<!--
+  This file is automatically @generated by {pkg_name} {pkg_version}
+  It is not intended for manual editing.
+-->
+
+<!DOCTYPE html>
+<html lang=\"en\">
+<head>
+  <meta charset=\"UTF-8\">
+  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">
+  <meta http-equiv=\"X-UA-Compatible\" content=\"ie=edge\">
+  <title>{module_name}</title>
+  <style>
+{minireset}
+
+{style}
+  </style>
+</head>
+<body>
+  <dl class=\"module\">
+",
+        module_name = "", // TODO: module name
+        minireset = MINIRESET.trim(),
+        style = STYLE.trim(),
+        pkg_name = pkg_name,
+        pkg_version = pkg_version,
+    )?;
+
+    for item in &module.items {
+        match item {
+            core::Item::Alias(alias) => compile_alias(&context, writer, alias, report)?,
+            core::Item::Struct(struct_ty) => {
+                compile_struct_ty(&context, writer, struct_ty, report)?
+            }
+        }
+    }
+
+    write!(
+        writer,
+        "\
+  </dl>
+</body>
+"
+    )?;
+
+    Ok(())
+}
+
+struct ModuleContext {
+    _file_id: FileId,
+}
+
+fn compile_alias(
+    context: &ModuleContext,
+    writer: &mut impl Write,
+    alias: &core::Alias,
+    report: &mut dyn FnMut(Diagnostic),
+) -> io::Result<()> {
+    writeln!(writer, "    <dt class=\"item alias\">")?;
+    writeln!(writer, "      <a href=\"#\">{}</a>", alias.name)?;
```

Comment:
I'm always a little wary of string interpolation into XML/HTML. Seems this is probably ok given identifiers are unlikely to contain characters that need escaping in HTML. May be worth considering a template language that guarantees a bit of correctness around tag generation and escaping at some point.

Reply:
Yeah, I was definitely thinking of going down that route in the future!